### ISSUE TITLE
Update Webhooks Extension to use Triggers 0.4

### ISSUE THREAD

**Issue by a-roberts** [tektoncd/experimental#493]
# Expected Behavior

Webhooks extension should work with Triggers 0.4

# Actual Behavior
It likely won't, lots of breaking changes so far - @dibyom @ncskier I'm sure I've seen a doc somewhere, can you provide a list of what's breaking so far please? I looked through the 0.3 release notes and didn't see anything deprecated but I don't think I'm imagining it

**Comment by ncskier** [comment]
Looking at what's on the [v0.4.0 Milestone](https://github.com/tektoncd/triggers/milestone/5) and what's been committed already, it looks like this will be the only breaking change:
- https://github.com/tektoncd/triggers/issues/332

We're planning on changing the APIGroup to `triggers.tekton.dev`. If I understand correctly, users will have to delete and recreate their Triggers resources with the new APIGroup when upgrading to the latest Triggers version after this change.

**Comment by a-roberts** [comment]
All good on Docker Desktop with 12 webhooks, having a look at OpenShift next with GitLab I'll deploy myself to be sure that's ok too, it'd be handy to have a big many webhook tester, pretty sure you've got a script somewhere @ncskier so wondering if we could make that more official. I'm using my own pipelines hotel branch at my fork with identical (except in name) pipeline/templates/bindings.

Note that as part of the testing I'm not taking into account displaying any new Triggers 0.4 features as that'd be Dashboard related, @ncskier are there any that spring to mind?

**Comment by ncskier** [comment]
> it'd be handy to have a big many webhook tester, pretty sure you've got a script somewhere @ncskier so wondering if we could make that more official.

My script was pretty rudimentary, and it didn't actually test triggering a large number of triggers (although you could easily modify it to do so). Here's the issue where I posted my script: https://github.com/tektoncd/triggers/issues/356#issuecomment-575777046. Since this script is pretty rudimentary, I don't really think it's worth publishing it/making it more official. However, we definitely should have something official for testing Triggers at a large scale. There's an existing issue for this here https://github.com/tektoncd/triggers/issues/406. If you're interested in working on testing Triggers at a large scale & getting performance metrics, feel free to contribute to this issue 👍 

> Note that as part of the testing I'm not taking into account displaying any new Triggers 0.4 features as that'd be Dashboard related, @ncskier are there any that spring to mind?

This one is already in `master`; it probably won't affect the Dashboard, but here it is just in case: https://github.com/tektoncd/triggers/pull/470.

This one hasn't been completed yet, so I don't know if it will make it into the v0.4.0 release or not, but it probably would affect the Dashboard: https://github.com/tektoncd/triggers/issues/371

This one will be a breaking change, but I'm pretty sure it won't affect the Dashboard (also it has not been implemented yet): https://github.com/tektoncd/triggers/issues/332.

There might be some additional features that get merged that are not under the Milestone right now, and I'll try to keep this issue updated in that case.

**Comment by a-roberts** [comment]
Triggers 0.4 has its first release candidate now so the next step is to get the pipelines in our hotel working, they're at https://github.com/pipeline-hotel/example-pipelines.git. I'll give this a go.

**Comment by a-roberts** [comment]
Putting this on hold while I look at https://github.com/tektoncd/experimental/issues/502

**Comment by a-roberts** [comment]
Current issue after moving the k8s dependencies up in our Gopkg.toml:

```
2020/04/03 15:34:04 Unexpected error running "go build": exit status 2
# github.com/tektoncd/experimental/webhooks-extension/pkg/endpoints
pkg/endpoints/webhook.go:934:39: cannot use client (type "github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/client-go/kubernetes/typed/certificates/v1beta1".CertificateSigningRequestInterface) as type context.Context in argument to csr.WaitForCertificate:
        "github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/client-go/kubernetes/typed/certificates/v1beta1".CertificateSigningRequestInterface does not implement context.Context (missing Deadline method)
pkg/endpoints/webhook.go:934:39: cannot use csrRecord (type *"github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/api/certificates/v1beta1".CertificateSigningRequest) as type "github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/client-go/kubernetes/typed/certificates/v1beta1".CertificateSigningRequestInterface in argument to csr.WaitForCertificate:
        *"github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/api/certificates/v1beta1".CertificateSigningRequest does not implement "github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/client-go/kubernetes/typed/certificates/v1beta1".CertificateSigningRequestInterface (missing Create method)
pkg/endpoints/webhook.go:934:63: cannot use 3600 * time.Second (type time.Duration) as type *"github.com/tektoncd/experimental/webhooks-extension/vendor/k8s.io/api/certificates/v1beta1".CertificateSigningRequest in argument to csr.WaitForCertificate
```

With @ncskier's help there are no actual Tekton related errors now in our code which so far sits at https://github.com/tektoncd/experimental/pull/503

(I dep ensure -v then ` kustomize build overlays/development | ko apply -f -`. Perhaps there's a new method we need to use or another dependency to pin.

@akihikokuroda FYI as it's to do with the certificate creation goodies you added a while back, which was all good using 

```
  name = "k8s.io/client-go"
  version = "kubernetes-1.13.4"
```

instead of

``` 
name = "k8s.io/client-go"
version = "v0.16.5"
```

🤔

**Comment by a-roberts** [comment]
Thanks @ncskier! Running with this now

**Comment by a-roberts** [comment]
https://github.com/tektoncd/experimental/issues/504 for the Pipeline Hotel changes, almost done making them

**Comment by a-roberts** [comment]
![image](https://user-images.githubusercontent.com/12814972/78565333-ac626f80-7815-11ea-813e-66b0a6b2eed4.png)

all done and working well on OCP 4.3 with Triggers 0.4 RC1, Pipelines 0.11, with my changes (so to both the example pipelines and our webhooks extension code - including RBAC, Go changes and a little JS).

I'll test scalability next making a dozen webhooks with various pipelines.

**Comment by a-roberts** [comment]
Nearly there, seeing

```
Checking pipelinerun buildah-pipeline-10-run-5qtn2 in namespace tekton-pipelines
Traceback (most recent call last):
  File "<stdin>", line 74, in <module>
ValueError: list.remove(x): x not in list
```

when i have lots of webhooks though. I'll address this. Not seeing any Triggers specific problems though.

**Comment by a-roberts** [comment]
Deletion isn't working either (at least in bulk) so have to fix this too.

**Comment by a-roberts** [comment]
Got it working a treat, happy with Triggers 0.4 @ncskier. I'm going to look into the issue @dibbles raised around deletion of webhooks with the same name next, and why an error is reported deleting many webhooks at once. 

I've also fixed a bug (it should be runsMissing.remove(missingDataEntry).

Then once this is merged and the dashboard changes are in, we can do a 0.6.1 release.

**Comment by a-roberts** [comment]
Merged so closing
